### PR TITLE
Sum over Idx can now detect the range, if available

### PR DESCRIPTION
**Pull request by Upabjojr**
Solving https://github.com/sympy/sympy/issues/9635

### CODE REVIEW

**Comment by Upabjojr** [comment]
@moorepants @bjodah @leosartaj @jcrist

**Comment by moorepants** [comment]
+1 if tests pass.

File: sympy/concrete/tests/test_sums_products.py [hunk 1]
```diff
@@ -833,7 +833,11 @@ def test_matrix_sum():
...
 
-def test_issue_9594():
+def test_indexed_idx_sum():
     i = symbols('i', cls=Idx)
     r = Indexed('r', i)
     Sum(r, (i, 0, 3)).doit() == [r.subs(i, j) for j in range(4)]
+
+    k = Idx('k', range=(1, 3))
+    A = IndexedBase('A')
+    Sum(A[k], k).doit() == [A[j] for j in range(1, 4)]
```
Comment (bjodah) [inline sympy/concrete/tests/test_sums_products.py]: I might be missing something obvious, but shouldn't there be an `assert` in front of the equality? (same question applies for the equality 4 lines above.

Comment (Upabjojr) [inline sympy/concrete/tests/test_sums_products.py]: You're right. I just edited a copy of the master branch version without checking.

**Comment by leosartaj** [comment]
We should clearly define the behavior in the following case.

Should the following be allowed?
```
>>> i = Idx('i', range=(1, 4))
>>> A = IndexedBase('A')
>>> Sum(A[i], (i, 1, 5).doit()
A[1] + A[2] + A[3] + A[4] + A[5]
```

Or this?
```
>>> Sum(A[i], (i, -1, 4)).doit()
>>> A[-1] + A[0] + A[1] + A[2] + A[3] + A[4]
```

@Upabjojr

File: sympy/concrete/tests/test_sums_products.py [hunk 2]
```diff
@@ -833,7 +833,11 @@ def test_matrix_sum():
...
-def test_issue_9594():
+def test_indexed_idx_sum():
     i = symbols('i', cls=Idx)
     r = Indexed('r', i)
-    Sum(r, (i, 0, 3)).doit() == [r.subs(i, j) for j in range(4)]
+    assert Sum(r, (i, 0, 3)).doit() == sum([Indexed('r', Idx(j)) for j in range(4)])
+
+    k = Idx('k', range=(1, 3))
+    A = IndexedBase('A')
+    assert Sum(A[k], k).doit() == sum([A[Idx(j, (1, 3))] for j in range(1, 4)])
```
Comment (bjodah) [inline sympy/concrete/tests/test_sums_products.py]: I don't quite understand why "A[Idx(j, (1, 3))]" is needed, wouldn't "A[j]" be enough? or maybe both should be tested?

**Comment by bjodah** [comment]
@leosartaj couldn't both be allowed?

I would like to suggest two additional tests:


```
def test_Indexed_sum():
    i = Idx('i', range=(1, 4))
    A = IndexedBase('A')
    s = Sum(A[i], i)
    assert s.doit() - A[1] - A[2] - A[3] - A[4] == 0
```

and

```
def test_Indexed_sum_complicated():
    j = Symbol('j', integer=True)
    i = Idx('i', range=(j-2, j+1))
    A = IndexedBase('A')
    s = Sum(A[i], i)
    assert s.doit() - A[j-2] - A[j-1] - A[j] - A[j+1] == 0
```

File: sympy/concrete/tests/test_sums_products.py (hunk 2 above)
Comment (Upabjojr) [inline sympy/concrete/tests/test_sums_products.py]:
```
In [6]: Sum(A[k], k).doit() == sum([A[Idx(j, (1, 3))] for j in range(1, 4)])
Out[6]: True

In [7]: Sum(A[k], k).doit() == sum([A[Idx(j)] for j in range(1, 4)])
Out[7]: False

In [8]: Sum(A[k], k).doit() == sum([A[j] for j in range(1, 4)])
Out[8]: False
```

It appears to be a bug. I would suggest that `Idx(number)` should return a _number_ (that is, no _Idx_ object) to solve this issue.

**Comment by Upabjojr** [comment]
```
In [9]: Idx(5)
Out[9]: 5

In [10]: Idx(5) == 5
Out[10]: False
```

Do we need this behaviour? _Idx_ is supposed to be a symbolic index, but with numbers, couldn't we just use numbers?

**Comment by Upabjojr** [comment]
> I would like to suggest two additional tests

@bjodah the first one is already tested, just written in another form. The second one could be added.

Anyway, your tests would not pass because `1 != Idx(1)` and, in the second test, very strange _args-expressions_ arise:
```
In [15]: srepr(s.doit().args[0])
Out[15]: "Indexed(IndexedBase(Symbol('A')), Idx(Symbol('j', integer=True), Tuple(Add(Symbol('j', integer=True), Integer(-2)), Add(Symbol('j', integer=True), Integer(1)))))"
```

**Comment by Upabjojr** [comment]
```
In [19]: Idx('i') == Idx('i', range=(1, 4))
Out[19]: False
```

_range_ is fundamental in the definition of _Idx_, equality-comparison is sensitive to it.

**Comment by bjodah** [comment]
@Upabjojr thank you for your explanations.
:+1: for making `Idx(5) == 5` return `True`

**Comment by leosartaj** [comment]
We need to think about the case when range for ``Idx`` is defined but ``Sum`` goes out of the range(see [this](https://github.com/sympy/sympy/pull/9694#issuecomment-122939860)). I don't think we should allow this behavior. It seems rather confusing to me. 

@Upabjojr @bjodah

**Comment by bjodah** [comment]
@leosartaj 
I failed to notice that the range was violated at first. Yes, then I agree that both those two cases should raise exceptions.

**Comment by Upabjojr** [comment]
A short summary of my opinion on index types:

* __Type 1__: a number represents the index, e.g. `A[5]`. In this case `Idx` is useless, I would change its constructor in order to return the number, i.e. `Idx(5)` return _5_.
* __Type 2__: a symbol standing for a number, e.g. `A[j]` where _j_ is meant to symbolize a unique number (no summation over _j_ can occur). _Idx_ objects are useless here, I would just keep symbols for this case. For this kind of index, expressions like `A[i] + A[j]` make sense. It may make sense to use an object for this kind of index if we were to associate the index with a basis and be able to handle changes of basis, but this is out of our current scope.
* __Type 3__: summation indices. The symbolic index may be summed. Expression like `A[i] + A[j]` __do not__ make any sense if _i_ and _j_ are abstract indices, because they are not summed over the same way.
* __Type 4__: abstract indices. Out of scope, let's ignore them.

My idea is that after `Sum.doit()`, there should be no _Idx_ indices, just symbols and/or numbers.

**Comment by Upabjojr** [comment]
> I would change its constructor in order to return the number, i.e. Idx(5) return 5.

This would also solve the test failure.

**Comment by bjodah** [comment]
@Upabjojr 
I agree with your point in __Type 1__

Bear with me if I misinterpreted you, but as for __Type 2__ I am not as convinced. Summation is not the only thing an index may signal. To me `Idx` would be equivalent with a Symbol carrying the assumption of being an integer, and optionally also carry bounds (lower and upper). Any implicit assumptions about summation is limiting the applicability of `Idx` in my humble opinion.

My same criticism then also applied to __Type 3__, if someone wants a sum, they can be explicit and type `Sum(A[i], i)` as you suggested in #9635 

I agree with your idea about `Sum.doit()` eliminating all `Idx` instances which the sum was performed over, i.e.:

```
>>> i = Idx('i', (0, 2))
>>> j = Idx('j', (0, 2))
>>> A = IndexedBase('A')
>>> Sum(A[i] + A[j], i).doit()
A[0] + A[1] + A[2] + 3*A[j]
```

**Comment by Upabjojr** [comment]
Apparently the current indices only support __Type 1__ and __Type 2__. The advantage of __Type 3__ is that you can eventually substitute your tensor expressions with arrays.

For example, if `A[i]` is associated to an array, and `B[i]` has another array (in this case one dimensional, but for `A[i, j, k, ...]` many be multi-dimensional), you can be sure that `A[i] + B[i]` is an array whose elements are the elementwise sum of the elements of the two original arrays. But in this case, what is `A[i] + B[k]`? This is not an array.

Unfortunately implementing __Type 3__ requires to modify _Add_ to check consistency. Yeah, __Type 3__ is out of scope here, that's what `sympy.tensor.tensor` implements.

**Comment by bjodah** [comment]
Since the constructor of Idx now returns integers when label is an integer we have a new (better - in my opinion) behavior. But this raises the question whether we should add a deprecation cycle? If anyone has acutally relied on `Idx(2, (1, 5)).lower` to work we will break their code. I think we may consider this a bug fix and hence skip the deprecation cycle, but the previous tests actually tested that very behavior. Git blame shows @jegerjensen as the author of those tests.

**Comment by Upabjojr** [comment]
> If anyone has acutally relied on Idx(2, (1, 5)).lower to work we will break their code

That sounds a bit preposterous. Why would you need to specify a range on a number? That doesn't make any sense at all. In SymPy we have assumptions on symbols (real, complex, positive, in the future may ranges), but no assumptions on number. Numbers already have all info about themselves.

By the way, what about monkey patching as a stopgap solution? That is, we return a _Number_ object, then we attach the methods after the object has been created. What do you think?

**Comment by bjodah** [comment]
Well, I actually think just returning an integer is the preferred way. But since there were actual tests of this I thought I should mention it if anyone could come up with a sane use case (I fail to see one).

**Comment by bjodah** [comment]
Just to make things clear: I really like this PR, and at this point I am :+1: to merging this.

**Comment by Upabjojr** [comment]
Wait, I thought a bit about the `Idx(5) != 5`. We don't need to make it equal. We could instead force `IndexedBase.__getitem__` to transform numbers to _Idx_ objects. That would solve the `A[5] != A[5]` problems and at the same time avoid possible compatibility breaks.

That is, all indices will be _Idx_ objects (could do the same for symbols).

@bjodah what do you think?

**Comment by bjodah** [comment]
I guess that would work. Intercepting in `IndexedBase.__getitem__` is a neat hack, but still a hack..

But you have convinced me `Idx(5, (1,3))` is silly (perhaps we should make it raise an excpetion), so making Idx(5) return just 5 sounds even better to me.

**Comment by Upabjojr** [comment]
OK, so this PR is good to merge. @moorepants

**Comment by bjodah** [comment]
I'll merge this in a few days if there are no objections.
BTW, `Product(A[i], i)` also works now (which is really nice), should we maybe have a test for it?

**Comment by Upabjojr** [comment]
> BTW, Product(A[i], i) also works now (which is really nice), should we maybe have a test for it?

@bjodah are you sure?

I got this:
```
In [1]: from sympy import *

In [2]: i = Idx('i')

In [3]: j = symbols('j', integer=True)

In [4]: A = IndexedBase('A')

In [5]: Product(A[j], (j, 1, 3)).doit()
Out[5]: A[1]*A[2]*A[3]

In [6]: Product(A[i], (i, 1, 3)).doit()
Out[6]: A[i]**3
```

It doesn't work on _Product_ if the index is an _Idx_ object.

**Comment by Upabjojr** [comment]
Now also _Product_ is supported. Any other issue?

**Comment by bjodah** [comment]
Brilliant! Thank you for your efforts, I'm merging this.
Will you add this to https://github.com/sympy/sympy/wiki/Release-Notes-for-0.7.7 ?

**Pull request merged**